docs/source/format/CanonicalExtensions.rst

```diff
@@ -283,6 +283,148 @@ UUID
    A specific UUID version is not required or guaranteed. This extension represents
    UUIDs as FixedSizeBinary(16) with big-endian notation and does not interpret the bytes in any way.
 
+Opaque
+=======
+
+Opaque represents a type or array that an Arrow-based system received from an
+external (often non-Arrow) system, which it cannot interpret or did not have
+support for in advance.  In this case, it can pass on Opaque to its clients to
+show that a field exists, but that it cannot interpret the field or data.
+
+Extension parameters:
+
+* Extension name: ``arrow.opaque``.
+
+* The storage type of this extension is any type.  If there is no underlying
+  data, the storage type should be Null.
+
+* Extension type parameters:
+
+  * **type_name** = the name of the unknown type in the external system.
+  * **vendor_name** = the name of the external system.
+
+* Description of the serialization:
+
+  A valid JSON object containing the parameters as fields.  In the future,
+  additional fields may be added, but all fields current and future are never
+  required to interpret the array.
+
+Rationale
+---------
+
+Arrow systems often wrap non-Arrow systems, and so they must be prepared to
+handle data types and data that don't have an equivalent Arrow type.  A client
+may still want to know of the existence of a field, or the types of other,
+supported fields.  So returning an error because of an unrecognized type in
+one column, or dropping unsupported fields/columns, are poor solutions.
+
+Of course, the Arrow system can use extension types.  But it will not have an
+extension type prepared for every possible type in advance; for example, the
+non-Arrow system may have its own extension mechanisms.  It could "make up" an
+extension type on the fly.  But this misleads clients who cannot tell if the
+type is truly supported or not by the intermediate Arrow application.
+
+The Opaque type can be used instead.  Because it explicitly means that the
+*intermediate* system does not support a type, it can be used to declare an
+unsupported field or column without silently losing data or erroring.  In
+other words: if an Arrow system encounters a non-Arrow type it was not
+prepared to handle, it can use Opaque to still pass the type on to a client.
+
+Applications **should not** make conventions around vendor_name and type_name.
```

Comment:
I'm not sure about this paragraph. In particular, it seems reasonable to make conventions around the vendor name.
Similarly, a reasonable convention for the type name would be something like "the original system's unabbreviated type name".

Reply:
That's very much not the point. This is not meant to replace having actual extension types, it is a stop gap as stated several times.

Reply:
Why does that preclude well-known vendor names?

Reply:
(1), I don't want to extend this document further. I thought it was already too long? :) (2), I don't want to encourage standardization here. It is explicitly meant as a free-for-all. As it states, it is meant for humans, and humans will not care about Postgres vs PostgreSQL.

Reply:
Well, things meant for humans tend to up parsed by machines anyway (think CLI interfaces).
That said, let's keep it like this if you feel so strongly about it.

Reply:
Perhaps a concrete example would help.

Reply:
This sentence
> Applications **should not** make conventions around vendor_name and type_name.

is vitally important for understanding what Opaque is intended for. I think it would be good to repeat this (or paraphrase it) higher above, immediately after defining the extension type parameters. This will prevent readers from immediately getting wrong ideas about what Opaque is intended for.

Perhaps we could paraphrase it like:

> Developers **should not** attempt to enable public semantic interoperability of Opaque by canonicalizing specific values of these parameters.

and add that just below where the parameters are defined.

Reply:
Thanks, updated.